### PR TITLE
Time Limited Credentials for SKUs

### PR DESCRIPTION
**Pull request by husobee**
### Summary

start of time limited credentials

### Type of change ( select one )

- [x] Product feature
- [ ] Bug fix
- [ ] Performance improvement
- [ ] Refactor
- [ ] Other

### Tested Environments

- [ ] Development
- [ ] Staging
- [ ] Production

### Before submitting this PR:

- [ ] Does your code build cleanly without any errors or warnings?
- [ ] Have you used auto closing keywords?
- [ ] Have you added tests for new functionality?
- [ ] Have validated query efficiency for new database queries?
- [ ] Have documented new functionality in README or in comments?
- [ ] Have you squashed all intermediate commits?
- [ ] Is there a clear title that explains what the PR does?
- [ ] Have you used intuitive function, variable and other naming?
- [ ] Have you requested security / privacy review if needed
- [ ] Have you performed a self review of this PR?

### CODE REVIEW

File: payment/controllers.go [hunk 1]
```diff
@@ -500,28 +503,101 @@ func GetOrderCreds(service *Service) handlers.AppHandler {
...
+			return handlers.WrapError(
+				err,
+				"failed to get order", http.StatusInternalServerError)
+		}
+
+		// look through order, find out what all the order item's credential types are
+		for i, v := range order.Items {
+			if i > 0 {
+				if v.CredentialType != credentialType {
+					// all the order items on the order need the same credential type
```
Comment (evq) [inline payment/controllers.go:524]: in the future we will likely need the ability to have mixed credential types on a single order

File: payment/controllers.go [hunk 2]
```diff
@@ -500,28 +503,101 @@ func GetOrderCreds(service *Service) handlers.AppHandler {
...
+					"only the Brave merchant can have time limited credentials", http.StatusBadRequest)
+			}
+			// is the order paid?
+			if !order.IsPaid() {
+				return handlers.WrapError(
+					fmt.Errorf("order is not paid"),
+					"cannot get credentials for unpaid orders", http.StatusBadRequest)
+			}
+
+			issuedAt := order.CreatedAt
```
Comment (evq) [inline payment/controllers.go:571]: we'll need to adapt this to use `last_paid_at`, right?

File: payment/controllers.go [hunk 3]
```diff
@@ -500,28 +503,101 @@ func GetOrderCreds(service *Service) handlers.AppHandler {
...
+			expiresAt := issuedAt.AddDate(0, 0, 35)
+
+			var credentials []TimeLimitedCreds
+			timeLimitedSecret := cryptography.NewTimeLimitedSecret([]byte(os.Getenv("BRAVE_MERCHANT_KEY")))
+
+			for _, item := range order.Items {
+				// iterate through order items, derive the time limited creds
+				timeBasedToken, err := timeLimitedSecret.Derive(issuedAt, expiresAt)
```
Comment (evq) [inline payment/controllers.go:579]: this doesn't appear to depend on the item, we should ensure time limited credentials for one item cannot be redeemed against another

File: payment/controllers.go [hunk 4]
```diff
@@ -500,28 +503,101 @@ func GetOrderCreds(service *Service) handlers.AppHandler {
...
+			for i := 0; i < len(*creds); i++ {
+				if (*creds)[i].SignedCreds == nil {
+					status = http.StatusAccepted
+					break
+				}
+			}
+
+			return handlers.RenderContent(r.Context(), creds, w, status)
+		} else if credentialType == timeLimited {
+			// only brave merchant
```
Comment (evq) [inline payment/controllers.go:558]: this is a lot of logic to have within the HTTP handler - we should consider abstracting some of it out for easier testing etc

File: payment/controllers.go [hunk 5]
```diff
@@ -500,28 +503,101 @@ func GetOrderCreds(service *Service) handlers.AppHandler {
...
+			}
+			// is the order paid?
+			if !order.IsPaid() {
+				return handlers.WrapError(
+					fmt.Errorf("order is not paid"),
+					"cannot get credentials for unpaid orders", http.StatusBadRequest)
+			}
+
+			issuedAt := order.CreatedAt
+			expiresAt := issuedAt.AddDate(0, 0, 35)
```
Comment (evq) [inline payment/controllers.go:572]: and expires at?

File: payment/controllers.go [hunk 6]
```diff
@@ -756,6 +832,47 @@ func VerifyCredential(service *Service) handlers.AppHandler {
...
+			issuedAt, err := time.Parse("2006-01-02", presentation.IssuedAt)
+			if err != nil {
+				return handlers.WrapError(err, "Error parsing issuedAt", http.StatusBadRequest)
+			}
+			expiresAt, err := time.Parse("2006-01-02", presentation.ExpiresAt)
+			if err != nil {
+				return handlers.WrapError(err, "Error parsing expiresAt", http.StatusBadRequest)
+			}
+
+			verified, err := timeLimitedSecret.Verify(issuedAt, expiresAt, presentation.Token)
```
Comment (evq) [inline payment/controllers.go:865]: this currently doesn't appear to check against the current time and we're not doing it here so is expiration actually enforced?

Comment (evq) [reply payment/controllers.go:865]: we should have a unit test that covers the expired case

**Review by evq — CHANGES_REQUESTED** [review]

File: payment/controllers.go (hunk 1 above)
Comment (husobee) [reply payment/controllers.go:524]: https://github.com/brave-intl/bat-go/issues/898

File: payment/controllers.go (hunk 5 above)
Comment (husobee) [reply payment/controllers.go]: merged and corrected: https://github.com/brave-intl/bat-go/pull/863/commits/4bd96fd8b29b50d3302ce21ddf2ee1adb3bbba05

File: payment/controllers.go (hunk 4 above)
Comment (husobee) [reply payment/controllers.go]: restructured: https://github.com/brave-intl/bat-go/pull/863/commits/4bd96fd8b29b50d3302ce21ddf2ee1adb3bbba05

File: payment/controllers.go (hunk 2 above)
Comment (husobee) [reply payment/controllers.go]: merged and corrected: https://github.com/brave-intl/bat-go/pull/863/commits/4bd96fd8b29b50d3302ce21ddf2ee1adb3bbba05

File: payment/controllers.go (hunk 6 above)
Comment (husobee) [reply payment/controllers.go:807]: checking expiry: https://github.com/brave-intl/bat-go/pull/863/commits/4bd96fd8b29b50d3302ce21ddf2ee1adb3bbba05 

tests coming next.

File: payment/controllers.go (hunk 3 above)
Comment (husobee) [reply payment/controllers.go]: fixed here https://github.com/brave-intl/bat-go/pull/863/commits/efe0e5b438741bbde13d803c3ba69ec54682aa72

File: payment/controllers.go (hunk 6 above)
Comment (husobee) [reply payment/controllers.go]: Test checking expiry here: https://github.com/brave-intl/bat-go/pull/863/commits/512b6945c26ab0643001acfffa7ca9b28f891a4f

File: payment/controllers.go [hunk 7]
```diff
@@ -779,6 +766,62 @@ func VerifyCredential(service *Service) handlers.AppHandler {
...
+
+			verified, err := timeLimitedSecret.Verify([]byte(presentation.Issuer), issuedAt, expiresAt, presentation.Token)
+			if err != nil {
+				return handlers.WrapError(err, "Error in token verification", http.StatusBadRequest)
+			}
+
+			if verified {
+				// check against expiration time
+				if time.Now().After(expiresAt) {
+					return handlers.RenderContent(r.Context(), "Credentials expired", w, http.StatusBadRequest)
```
Comment (orspetol) [inline payment/controllers.go:818]: Shouldn't this be a 403 (`http.StatusForbidden`)?

File: utils/cryptography/time_limited.go [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+// Verify - verify time limited credential based on date being bound within the expiration date of the credential
+func (secret TimeLimitedSecret) Verify(metadata []byte, date time.Time, expirationDate time.Time, token string) (bool, error) {
+	result, err := secret.Derive(metadata, date, expirationDate)
+	if err != nil {
+		return false, err
+	}
+	return subtle.ConstantTimeCompare([]byte(result), []byte(token)) == 1, nil
+}
+
+// NewTimeLimitedSecret - create a new time limited secret structuree
```
Comment (orspetol) [inline utils/cryptography/time_limited.go:34]: nit: misspell structure

File: payment/controllers.go (hunk 7 above)
Comment (husobee) [reply payment/controllers.go]: fixed: https://github.com/brave-intl/bat-go/pull/863/commits/3ee9f1c65a4b7df7629e9b69ac397d693d689eb0

File: utils/cryptography/time_limited.go (hunk 8 above)
Comment (husobee) [reply utils/cryptography/time_limited.go]: fixed: https://github.com/brave-intl/bat-go/pull/863/commits/3ee9f1c65a4b7df7629e9b69ac397d693d689eb0

**Review by evq — APPROVED** [review]

**Review by evq — APPROVED** [review]

**Pull request merged**